Client/Wallet/EthereumProviderHelper.swift

```diff
@@ -0,0 +1,171 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import WebKit
+import BraveCore
+import struct Shared.Logger
+
+private let log = Logger.browserLogger
+
+class EthereumProviderHelper: TabContentScript {
+  private static let supportedSingleArgMethods = [
+    "net_listening", "net_peerCount",
+    "net_version", "eth_chainId",
+    "eth_syncing", "eth_coinbase",
+    "eth_mining", "eth_hashrate",
+    "eth_accounts", "eth_newBlockFilter",
+    "eth_newPendingTransactionFilter"
+  ]
+  
+  private weak var tab: Tab?
+  
+  init(tab: Tab) {
+    self.tab = tab
+  }
+  
+  static func name() -> String {
+    return "walletEthereumProvider"
+  }
+  
+  func scriptMessageHandlerName() -> String? {
+    return "walletEthereumProvider_\(UserScriptManager.messageHandlerTokenString)"
+  }
+  
+  static func shouldInjectWalletProvider(_ completion: @escaping (Bool) -> Void) {
+    BraveWallet.KeyringServiceFactory.get(privateMode: false)?
+      .keyringInfo(BraveWallet.DefaultKeyringId, completion: { keyring in
+        completion(keyring.isKeyringCreated)
+      })
+  }
+  
+  private struct MessageBody: Decodable {
+    enum Method: String, Decodable {
+      case request
+      case isConnected
+      case enable
+      case send
+      case sendAsync
+      case isUnlocked
+    }
+    var securityToken: String
+    var method: Method
+    var args: String
+    
+    private enum CodingKeys: String, CodingKey {
+      case securityToken = "securitytoken"
+      case method
+      case args
+    }
+  }
+  
+  func userContentController(
+    _ userContentController: WKUserContentController,
+    didReceiveScriptMessage message: WKScriptMessage,
+    replyHandler: @escaping (Any?, String?) -> Void
+  ) {
+    guard let tab = tab,
+          !tab.isPrivate,
+          let provider = tab.walletProvider,
+          !message.frameInfo.securityOrigin.host.isEmpty, // Fail if there is no last committed URL yet
+          message.frameInfo.isMainFrame, // Fail the request came from 3p origin
+          JSONSerialization.isValidJSONObject(message.body),
+          let messageData = try? JSONSerialization.data(withJSONObject: message.body, options: []),
+          let body = try? JSONDecoder().decode(MessageBody.self, from: messageData),
+          body.securityToken == UserScriptManager.securityTokenString
+    else {
+      log.error("Failed to handle ethereum provider communication")
+      return
+    }
+    
+    // The web page has communicated with `window.ethereum`, so we should show the wallet icon
+    tab.isWalletIconVisible = true
+    
+    let securityOrigin = message.frameInfo.securityOrigin
+    let origin = "\(securityOrigin.protocol)://\(securityOrigin.host):\(securityOrigin.port)"
```

Comment:
Are we checking the origins after the message is sent, where?

Reply:
At this point the origin is sent directly into brave-core and handled via the provider implementation https://github.com/brave/brave-core/blob/31c31e76c8899ca71b0314c087ed134da065bd48/components/brave_wallet/browser/brave_wallet_provider_impl.cc#L831

After that it may or may not call `ShowPanel` and we simply fetch pending requests in a particular order, but it isn't tied to a specific tab

Reply:
What happens if the communication is started from an iframe? Is it blocked or does it start from the tab page?

Reply:
The communication JS & ethereal provider from brave-core  is only injected into main frame, so `window.ethereum` will not exist on iframes.